crates/fuzzing/src/oracles/engine.rs

```diff
@@ -0,0 +1,145 @@
+//! Define the interface for differential evaluation of Wasm functions.
+
+use crate::generators::{DiffValue, ModuleFeatures};
+use crate::oracles::{diff_wasmi::WasmiEngine, diff_wasmtime::WasmtimeEngine};
+use anyhow::{bail, Ok};
+use arbitrary::Unstructured;
+use std::collections::hash_map::DefaultHasher;
+
+/// Pick one of the engines implemented in this module that is compatible with
+/// the Wasm features passed in `features` and, when fuzzing Wasmtime against
+/// itself, an existing `wasmtime_engine`.
+pub fn choose(
+    u: &mut Unstructured<'_>,
+    features: &ModuleFeatures,
+    wasmtime_engine: &WasmtimeEngine,
+) -> anyhow::Result<Box<dyn DiffEngine>> {
+    // Filter out any engines that cannot match the given configuration.
+    let mut engines: Vec<Box<dyn DiffEngine>> = vec![];
+    if let Result::Ok(e) = WasmtimeEngine::arbitrary_with_compatible_config(u, wasmtime_engine) {
+        engines.push(e)
+    }
+    if let Result::Ok(e) = WasmiEngine::new(features) {
+        engines.push(e)
+    }
+    #[cfg(feature = "fuzz-spec-interpreter")]
+    if let Result::Ok(e) = crate::oracles::diff_spec::SpecInterpreter::new(features) {
+        engines.push(e)
+    }
+
+    // Choose one of the remaining engines.
+    if !engines.is_empty() {
+        let index: usize = u.int_in_range(0..=engines.len() - 1)?;
+        Ok(engines.swap_remove(index))
+    } else {
+        bail!("no engines to pick from")
+    }
+}
+
+/// Provide a way to instantiate Wasm modules.
+pub trait DiffEngine {
+    /// Create a new instance with the given engine.
+    fn instantiate(&self, wasm: &[u8]) -> anyhow::Result<Box<dyn DiffInstance>>;
+}
+
+/// Provide a way to evaluate Wasm functions--a Wasm instance implemented by a
+/// specific engine (i.e., compiler or interpreter).
+pub trait DiffInstance {
+    /// Return the name of the engine behind this instance.
+    fn name(&self) -> &'static str;
+
+    /// Evaluate an exported function with the given values.
+    fn evaluate(
+        &mut self,
+        function_name: &str,
+        arguments: &[DiffValue],
+    ) -> anyhow::Result<Vec<DiffValue>>;
+
+    /// Check if instances of this kind are actually hashable--not all engines
+    /// support this.
+    fn is_hashable(&self) -> bool;
+
+    /// If the instance `is_hashable()`, this method will try to hash the
+    /// following exported items in the instance: globals, memory.
+    ///
+    /// TODO allow more types of hashers.
+    fn hash(&self, state: &mut DefaultHasher) -> anyhow::Result<()>;
```

Comment:
Reading over the implementations of this could this trait perhaps get changed to have something different like "get the hashable export named `foo`" to avoid requiring hardcoding the hash algorithm between all engines? That would also allow more direct comparisons and better errors along the lines of "the state of this global/memory is different" rather than "some state is different"

Reply:
Yeah, this probably does need some changes but I'm not yet entirely clear what they are; I think looking at what is necessary for the OCaml spec interpreter might make things more clear (as mentioned [here](https://github.com/bytecodealliance/wasmtime/pull/4515#issuecomment-1212486350)).